### PR TITLE
Interval Lock utility

### PR DESCRIPTION
**Pull request by cmc333333**
Use openresty's shared memory (with expiration) functionality to implement
time-based locks. Function calls can be wrapped so that they are only executed
once in a given period of time. Rather than requiring functions to track their
own locked status, the onus is placed on the caller, allowing the wrapped
function to be called explicitly if needed or placed behind multiple locks.

Resolves https://github.com/18F/api.data.gov/issues/292

### CODE REVIEW

File: src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua [hunk 1]
```diff
@@ -89,7 +78,10 @@ local function check(premature)
     return
   end
 
-  local ok, err = pcall(do_check)
+  -- here we subtract the lock expiration time by 1ms to prevent
+  -- a race condition with the next timer event.
+  local ok, err = interval_lock('distributed-last-pull', delay - 0.001,
```
Comment (GUI) [inline src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua]: It looks like we end up subtracting 2ms to handle the race condition, since this subtraction is happening here and also inside the `interval_lock` function. Not a huge deal, but I think it might make the most sense to only subtract 1ms inside `interval_lock` and shift the comment explaining the 1ms funkiness into the `interval_lock` function. That way the caller doesn't really have to be aware of this issue.

**Comment by GUI** [comment]
Sweet, thanks for tackling this!

This is looking good, my main bit of feedback is that I think we still need to keep the `resty.lock` usage in addition to this new interval lock. The reason for both types of locks is that they do slightly different things. The `resty.lock` prevents the same function from executing concurrently, since it requires an explicit lock and unlock when the task finishes. In contrast, the interval lock ensures the function only executes once per interval, but it isn't actually aware of when the task finishes (since it's just a clock-based timer that expires). For tasks that take less than the interval to execute, then the interval lock is all that's really needed, but to account for the possibility that a task unexpectedly takes longer than the interval, that's why I think it would be good to keep the `resty.lock` usage also in place. Does that make sense? I know all these timers and concurrent workers gets a bit confusing, so I can try to diagram out why I think we need both types of locks in place still. But if you have any other ideas on tackling any of this, just give a shout.

Then I think the only other thing is that once we settle on an approach, we'll want to update the other background jobs that are doing very similar things with locking. I think this is the full list:

- [`src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua)
- [`src/api-umbrella/proxy/jobs/elasticsearch_setup.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/elasticsearch_setup.lua)
- [`src/api-umbrella/proxy/jobs/load_api_users.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/load_api_users.lua)
- [`src/api-umbrella/proxy/jobs/load_db_config.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/load_db_config.lua)
- [`src/api-umbrella/proxy/jobs/resolve_backend_dns.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/resolve_backend_dns.lua)
- [`src/api-umbrella/proxy/jobs/distributed_rate_limit_pusher.lua`](https://github.com/NREL/api-umbrella/blob/master/src/api-umbrella/proxy/jobs/distributed_rate_limit_pusher.lua) (this one actually seems to be missing any type of locking, so that might be an initial oversight on my part)

**Comment by GUI** [comment]
And actually, in looking at all those job tasks again, I'm inspired by your abstraction of the `interval_lock` function and wondering if we could better abstract the bigger pattern of repeating a function call once per interval with all the proper locks. Because there's definitely some commonality with the `check`, `do_check` pattern between all the functions, and if we also need `resty.lock` in place surrounding all these, it seems like that's a fair amount of duplicate boiler-plate across all these.

Here's some quick, ugly pseudo-code of some initial thoughts:

```lua
-- 
-- Example job file wanting to call "do_check" once per some interval.
--

local interval = 0.25

function do_check()
  -- Do stuff.
end

-- No more "check" wrapper method needed inside each job file.

function spawn()
  new_timer(0, repeat_task, "distributed_rate_limit_puller", interval, do_check)
end



--
-- Example repeat_task library
--

function repeat_task_lock_wrapper(name, fn)
  -- Interval lock - Ensure "fn" only gets called once per interval.
  ok, err = ngx.shared.locks:add(name .. "-interval", true, interval - 0.001)
  if not ok then
    return
  end

  -- Concurrency lock - Ensure "fn" never gets called if it's already running
  -- (and taking longer than interval).
  repeat_lock = lock:new("locks", { ["timeout"] = 0 })
  lock_err = repeat_lock:lock(name)
  if lock_err then
    return
  end

  -- Call the real function we want to call once per interval.
  pcall(fn)

  -- Unlock the concurrency lock.
  repeat_lock:unlock()
end

function repeat_task(premature, name, interval, fn)
  -- Return if nginx is shutting down.
  if premature then return end

  -- Call the underlying function with locks to ensure it only gets called once per interval.
  ok, err = pcall(repeat_task_lock_wrapper, name, fn)

  -- Regardless of success, schedule another attempt in interval.
  ok, err = new_timer(interval, repeat_task, name, interval, fn)
end
```

Any thoughts on this kind of further abstraction?

File: src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua (hunk 1 above)
Comment (cmc333333) [inline src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua]: Whoopsie; last minute edit fail. The intent was originally the reverse -- why would `interval_lock` know about a timer? It's a safe play, but I think it assumes too much knowledge about it's calling context.

Given that we're expecting to combine this sort of lock with a timer often, perhaps it makes more sense to add a related function which also sets a timer for callbacks?

**Comment by cmc333333** [comment]
Hah; same wavelength re: abstracting timers.

I understand now about `resty.lock`, thanks!

**Comment by cmc333333** [comment]
@GUI Updated with the more comprehensive timer approach. Please double check the logic for me -- I'm pretty sure everything composes, but as you noted, this is hard stuff.

If it's all good, I'll follow up with commits for the other timers.

**Comment by GUI** [comment]
Sorry again for the delay, but I've finally had a chance to look at this. This looks really solid and super nice in how it can cleanup all of our similar background tasks. :sparkling_heart:

As we've both realized, this stuff is tricky, and these are some dense functions, but after stepping through it all, everything makes sense and looks good. And in playing around with it locally, I've verified the underlying function is only getting called once per interval, right on schedule.

I think my only small items of feedback are to add some additional error logging in a few places I spotted in case things unexpectedly blow up. I'll add some inline comments to this pull request (and sorry to continue to quibble on super tiny things). But this looks great, and if you want to go ahead and apply this to the other background tasks in the app, that would be dandy. Or we can also merge this in and I'd be happy to update the other background tasks, whichever you'd like to do.

File: src/api-umbrella/utils/interval_lock.lua [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param interval - the length of time until expiry (seconds)
+-- @param fn - function to execute within the interval
+-- @return ok, error
+_M.timeout_exec = function(name, interval, fn)
+  local ok, err = ngx.shared.interval_locks:add(name, true, interval)
+  if not ok and err == "exists" then
+    return true, nil
+  elseif not ok then
+    return false, err
```
Comment (GUI) [inline src/api-umbrella/utils/interval_lock.lua]: Can we add an `ngx.log(ngx.ERR)` statement inside here before returning false? This is a case that really shouldn't happen, but if it does, it likely means something something bad has happened like we've run out space on the `interval_locks` shared memory table. So I think logging an error would be helpful so we at least know if we've hit this unexpected situation.

File: src/api-umbrella/utils/interval_lock.lua [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+-- @param name - a unique identifier for this lock (automatically namespaced)
+-- @param fn - function to execute if the lock can be held
+-- @return ok, error
+_M.mutex_exec = function(name, fn)
+  local check_lock = lock:new("locks", {["timeout"] = 0})
+  local _, lock_err = check_lock:lock("mutex_exec:" .. name)
+  if lock_err then
+    return true, nil
+  else
+    local results = pcall(fn)
```
Comment (GUI) [inline src/api-umbrella/utils/interval_lock.lua]: What are your thoughts about checking the status of the `pcall` and logging an error here if it didn't succeed? `pcall` is used to call functions and capture any errors they might raise. But since the errors don't bubble up further it means that unless you check the status of the `pcall` result and do something with it, errors will silently be swallowed. So anytime I use pcall, I usually check the result immediately afterwards and log if there was an error. Here's one example from the previous usage: https://github.com/NREL/api-umbrella/blob/3920e2852ae4ff074c205f606462050fd4310ea2/src/api-umbrella/proxy/jobs/distributed_rate_limit_puller.lua#L92-L95

Also, just FYI, the return signature on `pcall` is a bit funny. The first variable it returns will always be the status of whether the call succeeded. If the call failed, then the second variable will be the error message about what failed. If the call succeeded, then the second variable will be the original return value of the function you were calling (in our cases, I'm not sure we have a need for the real return value from these background jobs, since I'm not sure they return anything, but I suppose that could change at some point).

File: src/api-umbrella/utils/interval_lock.lua [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+-- @param fn - function to execute within the interval
+-- @return ok, error
+_M.timeout_exec = function(name, interval, fn)
+  local ok, err = ngx.shared.interval_locks:add(name, true, interval)
+  if not ok and err == "exists" then
+    return true, nil
+  elseif not ok then
+    return false, err
+  else
+    return pcall(fn)
```
Comment (GUI) [inline src/api-umbrella/utils/interval_lock.lua]: Same as https://github.com/NREL/api-umbrella/pull/191/files#r45168023 I might be inclined to check the status of this `pcall` and log failures. However, this one is probably the more important one to log, since this is the one that's actually going to be calling our target function in the app, where I'm assuming it's more likely things could unexpectedly break. The other pcalls in this library would just break if something in this library unexpectedly broke, so it seems less likely, but logging failures for all of these still might not be a bad idea.

File: src/api-umbrella/utils/interval_lock.lua [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+_M.repeat_exec = function(interval, fn)
+  -- schedule the next call
+  local ok, err = ngx.timer.at(interval, function()
+    _M.repeat_exec(interval, fn)
+  end)
+  if not ok and err ~= "process exiting" then
+    ngx.log(ngx.ERR, "failed to create timer: ", err)
+  else
+    -- execute fn now (may tie down this thread)
+    pcall(fn)
```
Comment (GUI) [inline src/api-umbrella/utils/interval_lock.lua]: Same as https://github.com/NREL/api-umbrella/pull/191/files#r45168425 We might consider logging `pcall` errors.

File: src/api-umbrella/utils/interval_lock.lua [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+--- Call a function every `interval` amount of time
+-- Set a timeout for next execution and run the provided function. Note that,
+-- should the function execution take more time than the `interval`, this will
+-- quickly backup. Combine with the mutexes above via `repeat_with_mutex`
+-- @param interval - the length of time until repeated (seconds)
+-- @param fn - function to execute when the next timeout occurs
+_M.repeat_exec = function(interval, fn)
+  -- schedule the next call
+  local ok, err = ngx.timer.at(interval, function()
```
Comment (GUI) [inline src/api-umbrella/utils/interval_lock.lua]: The function that `ngx.timer.at` calls always has a first argument of `premature`: https://github.com/openresty/lua-nginx-module#ngxtimerat I'm not exactly sure how relevant it is in our case, but I might change things around to something like:

```
function(premature)
  if premature then
    return
  end
  ...
```

`premature` gets set to true if the timer gets called while the nginx process is trying to exit as part of a shutdown or SIGHUP reload. So again, I'm not totally sure of the real impact of this, but I think it may help the reload or shutdown to happen more quickly and gracefully since it prevents another round of the background jobs from beginning to execute during this shutdown (and in general, I've seen this `if premature then return` logic in lots of other openresty examples).

File: src/api-umbrella/utils/interval_lock.lua (hunk 2 above)
Comment (cmc333333) [inline src/api-umbrella/utils/interval_lock.lua]: :+1: will add

File: src/api-umbrella/utils/interval_lock.lua (hunk 3 above)
Comment (cmc333333) [inline src/api-umbrella/utils/interval_lock.lua]: Logging the `pcall` results makes a lot of sense; adding. Thinking through the calls further, it doesn't seem like the results could be used meaningfully (as they're tied to a timer, outside of the calling thread), so I'll go ahead and remove them.

File: src/api-umbrella/utils/interval_lock.lua (hunk 4 above)
Comment (cmc333333) [inline src/api-umbrella/utils/interval_lock.lua]: :+1:

File: src/api-umbrella/utils/interval_lock.lua (hunk 5 above)
Comment (cmc333333) [inline src/api-umbrella/utils/interval_lock.lua]: :+1:

File: src/api-umbrella/utils/interval_lock.lua (hunk 6 above)
Comment (cmc333333) [inline src/api-umbrella/utils/interval_lock.lua]: :+1: I was hoping for exactly this type of feedback, thank you

**Comment by cmc333333** [comment]
@GUI updated. Would you mind double checking the lock in `load_db_config`? I think it's okay to move the check (implicitly) before `WORKER_GROUP_ID`, but would appreciate confirmation.

CI is failing because I pulled in master, which needs `dtrace`. I can add installing `systemtap-sdt-devel` to the CI script?

**Comment by GUI** [comment]
Ah, oops, sorry for the dtrace issue. I actually just removed that since it was causing build issues on other distros and I don't think we really need it: https://github.com/NREL/api-umbrella/commit/1665e065b3c01d4b4c45fc4facc6d52ae0d68a78

I also should have caught this sooner, but the CI environment on master was using a cached copy of the build prior to introducing this dependency. I've attempted to fix this moving forward by giving us a way to clear the build cache when we introduce these type of build changes: https://github.com/NREL/api-umbrella/commit/518efe9b529193cdc69c1b944530ba328053122d

So in short, I don't think you need to worry about the dtrace build issue, I can just verify when merging this back into master.

I think this look good, but I'll give this a more thorough look and see about the `WORKER_GROUP_ID ` later today. Thanks!

**Comment by GUI** [comment]
Looks good. Thanks again! I'll get this deployed tomorrow.

**Comment by GUI** [comment]
Just FYI, I had to tweak two things I didn't catch earlier, which I've tried to explain in more detail in the commits: https://github.com/NREL/api-umbrella/commit/94b601cbbaf3ac89a92aff3b72e273f1f9ee72f2 & https://github.com/NREL/api-umbrella/commit/0a24560291d9a6720d6bab52fc57bd2db791aad4

But otherwise, everything looks good. I've deployed this to production, and we've seen a nice decrease in the MongoDB CPU utilization (since this reduces a lot of the redundant and simultaneous polling). Yay! :+1: 

screen shot 2015-11-24 at 1 06 57 pm
